pkg/cloudprovider/provider/gce/service.go

```diff
@@ -0,0 +1,129 @@
+//
+// Google Cloud Provider for the Machine Controller
+//
+
+package gce
+
+//-----
+// Imports
+//-----
+
+import (
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/oauth2"
+	"google.golang.org/api/compute/v1"
+)
+
+//-----
+// Constants
+//-----
+
+const (
+	// Timeouts.
+	timeoutNormal = 10 * time.Minute
+)
+
+// Google compute operation status.
+const (
+	statusDone = "DONE"
+)
+
+//-----
+// Service
+//-----
+
+// service wraps a GCE compute service for the extension with helper methods.
+type service struct {
+	*compute.Service
+}
+
+// connectComputeService establishes a service connection to the Compute Engine.
+func connectComputeService(cfg *config) (*service, error) {
+	svc, err := compute.New(cfg.jwtConfig.Client(oauth2.NoContext))
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to Google Cloud: %v", err)
+	}
+	return &service{svc}, nil
+}
+
+// networkInterfaces returns the configured network interfaces for an instance creation.
+func (svc *service) networkInterfaces(cfg *config) ([]*compute.NetworkInterface, error) {
+	ifc := &compute.NetworkInterface{
+		AccessConfigs: []*compute.AccessConfig{
+			{
+				Name: "External NAT",
+				Type: "ONE_TO_ONE_NAT",
+			},
+		},
+		Network: "global/networks/default",
+	}
+	return []*compute.NetworkInterface{ifc}, nil
+}
+
+// attachedDisks returns the configured attached disks for an instance creation.
+func (svc *service) attachedDisks(cfg *config) ([]*compute.AttachedDisk, error) {
+	sourceImage, err := cfg.sourceImageDescriptor()
+	if err != nil {
+		return nil, err
+	}
+	bootDisk := &compute.AttachedDisk{
+		Boot:       true,
+		AutoDelete: true,
+		InitializeParams: &compute.AttachedDiskInitializeParams{
+			DiskSizeGb:  cfg.diskSize,
+			DiskType:    cfg.diskTypeDescriptor(),
+			SourceImage: sourceImage,
+		},
+	}
+	return []*compute.AttachedDisk{bootDisk}, nil
+}
+
+// refreshOperation requests a fresh copy of the passed operation containing
+// the updated status.
+func (svc *service) refreshOperation(projectID string, op *compute.Operation) (*compute.Operation, error) {
+	switch {
+	case op.Zone != "":
+		zone := path.Base(op.Zone)
+		return svc.ZoneOperations.Get(projectID, zone, op.Name).Do()
+	case op.Region != "":
+		region := path.Base(op.Region)
+		return svc.RegionOperations.Get(projectID, region, op.Name).Do()
+	default:
+		return svc.GlobalOperations.Get(projectID, op.Name).Do()
+	}
+}
+
+// waitOperation waits for a GCE operation to be completed or timed out.
+func (svc *service) waitOperation(projectID string, op *compute.Operation, timeout time.Duration) (err error) {
+	started := time.Now()
+	waiting := 100 * time.Millisecond
```

Comment:
Instead if building you own polling function i would recommend to use something which already exists:
https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L306

Reply:
Nice, did not know. Mine has been only stolen. 😄